### DISCUSSION TITLE
Using Extract Geometry on a Complex PolyData Object Causes it to Calculate Volume Incorrectly

### DISCUSSION THREAD

**Discussion by Connor56** [pyvista/pyvista#1936] [Q&A]
I have been doing some work recently which requires me to cut slabs from crystal grains on a computer. Essentially this means I have a set of points, which can be used to produce an outer surface with a polyhedral type shape.

Due to the way the points are manipulated before being used in PyVista some of the points are negative, and the point system does not touch [0, 0, 0]. When I calculate the volume of these points I get far larger values than one would expect. Is this due to the point system not being centred around [0, 0, 0]?

I tested this for negative points and it made no difference, so I'm assuming that's the issue?

Has anyone ran into something similar before? Is there an explanation/ reason for this behaviour, or is it a bug?

**Comment by banesullivan** [comment]
Can you share an example mesh?

**Reply by Connor56** [reply]
Apologies for not responding for so long, I'm currently under pressure to finish some work so don't have the time to recreate the issue. I will do in the next 3-4 weeks I promise!

**Comment by wangjx520** [comment]
I have also found this problem, pyvista does not correctly calculate the volume of meshes above the octahedron in my experiments. My solution was to use scipy.spatial.ConvexHull(points).volume. i verified it on an octahedral mesh and it worked.

**Reply by MatthewFlamm** [reply]
Can you share a minimal reproducible example?

This would help us test and find a fix.

**Reply by wangjx520** [reply]
Of course, if it helps to improve pyvista, I will be very happy!
```
import pyvista as pv
ugrid=pv.read('exp.vtu')
ugrid.plot(notebook=True,show_edges=True)
ugrid.bounds
print(ugrid.cell_centers().points) #out of the bounds
ugrid.get_cell(4).plot()
ugrid['Volume']=ugrid.compute_cell_sizes(length=False,area=False,volume=True)['Volume'] #some units have a calculated volume value of 0
print('The total volume of all cells calculated by pyvista is',ugrid['Volume'].sum())

import numpy as np
from scipy.spatial import ConvexHull

def calculate_volume(points):
    """
    通过给定的顶点坐标计算一个凸多面体的体积。
    :param points: 顶点坐标的列表，形状为 (n_points, 3)
    :return: 体积
    """
    hull = ConvexHull(points)
    return hull.volume

volumes=[]
for cellid in range(ugrid.n_cells):
    points=ugrid.get_cell(cellid).points[:,:3]
    volumes.append(calculate_volume(points))
print('The total volume of all cells calculated by scipy is',np.array(volumes).sum())
```
[exp.zip](https://github.com/user-attachments/files/16276431/exp.zip)

Perhaps, because my vtu file is not legal? This is a voronoi diagram I converted to UnstructuredGrid, if possible, please help me check if the vtu is correct!

**Comment by MatthewFlamm** [comment]
pyvista uses [`vtkMassProperties`](https://vtk.org/doc/nightly/html/classvtkMassProperties.html#details) under the hood.  It says

> Warning
> Currently only triangles are processed.

pyvista should document this.

**Reply by MatthewFlamm** [reply]
Furthermore, a `PolyData` can contain several disconnected closed surfaces, this also won't work.  They need to be separated first.  This all needs to be documented.

**Reply by MatthewFlamm** [reply]
The mesh is already triangulated under the hood:

https://github.com/pyvista/pyvista/blob/a98885564ae6bbb8a58f98c5ca27d3b97b142cc4/pyvista/core/pointset.py#L1528-L1530

I wonder if we also need to compute the surface normals.  It isn't explicitly stated in the vtk documentation, but this is a common need for PolyData operations.

There is this note in the [vtk documentation](https://vtk.org/doc/nightly/html/classvtkMassProperties.html#a0930a813734a344f09819a269371a40f), which gives a hint that it might be necessary:

```
Either the PolyData is not closed
Or the PolyData contains triangle that are flipped
```

**Reply by banesullivan** [reply]
Indeed this needs to be documented. Though some of it is already done for us in this example with `split_bodies ()` https://docs.pyvista.org/version/stable/examples/01-filter/compute-volume#a-real-dataset
